File: apps/macos/Sources/SloerShot/OcrService.swift
import Foundation
import Vision
import CoreGraphics

// Real on-device OCR via Apple Vision. Produces the OcrResult JSON that shotcore consumes,
// so all region / reading-order / clipboard formatting runs in the tested core
// (ShotCore.ocrTextInRegion / ocrSingleLine). Vision only supplies recognized lines + boxes.
enum OcrService {
 static func recognizeJSON(cgImage: CGImage, width: Int, height: Int) -> String? {
 let request = VNRecognizeTextRequest()
 request.recognitionLevel = .accurate
 let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
 do {
 try handler.perform([request])
 } catch {
 return nil
 }
 guard let observations = request.results else { return nil }
 let w = Double(width)
 let h = Double(height)
 var lines: [[String: Any]] = []
 for obs in observations {
 guard let candidate = obs.topCandidates(1).first else { continue }
 // Vision boxes are normalized, origin bottom-left; convert to top-left pixels.
 let bb = obs.boundingBox
 let bbox: [String: Any] = ["x": bb.minX * w, "y": (1.0 - bb.maxY) * h, "w": bb.width * w, "h": bb.height * h]
 let word: [String: Any] = ["text": candidate.string, "bbox": bbox, "confidence": Double(candidate.confidence)]
 lines.append(["text": candidate.string, "bbox": bbox, "words": [word]])
 }
 guard let data = try? JSONSerialization.data(withJSONObject: ["lines": lines]) else { return nil }
 return String(data: data, encoding: .utf8)
 }
}

File: apps/windows/SloerShot.App/Interop/OcrService.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Windows.Graphics.Imaging;
using Windows.Media.Ocr;

namespace SloerShot.Interop;

// Real on-device OCR via Windows.Media.Ocr. Produces the OcrResult JSON that shotcore
// consumes, so all region/reading-order/clipboard formatting runs in the tested core
// (ShotCore.OcrTextInRegion / OcrSingleLine). The OS engine only supplies words + boxes.
public static class OcrService
{
 public static bool IsAvailable => OcrEngine.AvailableRecognizerLanguages.Count > 0;

 public static async Task<string?> RecognizeJsonAsync(SoftwareBitmap bitmap)
 {
 var engine = OcrEngine.TryCreateFromUserProfileLanguages();
 if (engine == null) return null;
 var result = await engine.RecognizeAsync(bitmap);
 var lines = new List<object>();
 foreach (var line in result.Lines)
 {
 var words = new List<object>();
 double minX = double.MaxValue, minY = double.MaxValue, maxX = 0, maxY = 0;
 foreach (var word in line.Words)
 {
 var r = word.BoundingRect;
 words.Add(new { text = word.Text, bbox = new { x = r.X, y = r.Y, w = r.Width, h = r.Height }, confidence = 1.0 });
 minX = Math.Min(minX, r.X);
 minY = Math.Min(minY, r.Y);
 maxX = Math.Max(maxX, r.X + r.Width);
 maxY = Math.Max(maxY, r.Y + r.Height);
 }
 if (line.Words.Count == 0) { minX = 0; minY = 0; maxX = 0; maxY = 0; }
 lines.Add(new { text = line.Text, bbox = new { x = minX, y = minY, w = maxX - minX, h = maxY - minY }, words });
 }
 return JsonSerializer.Serialize(new { lines });
 }

 public static async Task<string?> RecognizeTextAsync(SoftwareBitmap bitmap)
 {
 var engine = OcrEngine.TryCreateFromUserProfileLanguages();
 if (engine == null) return null;
 var result = await engine.RecognizeAsync(bitmap);
 return result.Text;
 }
}
